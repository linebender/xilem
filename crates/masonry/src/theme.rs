// This software is licensed under Apache License 2.0 and distributed on an
// "as-is" basis without warranties of any kind. See the LICENSE file for
// details.

//! Theme keys and initial values.

#![allow(missing_docs)]

use vello::peniko::Color;

use crate::Insets;

// Colors are from https://sashat.me/2017/01/11/list-of-20-simple-distinct-colors/
// They're picked for visual distinction and accessbility (99 percent)

pub const WINDOW_BACKGROUND_COLOR: Color = Color::rgb8(0x29, 0x29, 0x29);
pub const TEXT_COLOR: Color = Color::rgb8(0xf0, 0xf0, 0xea);
pub const DISABLED_TEXT_COLOR: Color = Color::rgb8(0xa0, 0xa0, 0x9a);
pub const PLACEHOLDER_COLOR: Color = Color::rgb8(0x80, 0x80, 0x80);
pub const PRIMARY_LIGHT: Color = Color::rgb8(0x5c, 0xc4, 0xff);
pub const PRIMARY_DARK: Color = Color::rgb8(0x00, 0x8d, 0xdd);
pub const PROGRESS_BAR_RADIUS: f64 = 4.;
pub const BACKGROUND_LIGHT: Color = Color::rgb8(0x3a, 0x3a, 0x3a);
pub const BACKGROUND_DARK: Color = Color::rgb8(0x31, 0x31, 0x31);
pub const FOREGROUND_LIGHT: Color = Color::rgb8(0xf9, 0xf9, 0xf9);
pub const FOREGROUND_DARK: Color = Color::rgb8(0xbf, 0xbf, 0xbf);
pub const DISABLED_FOREGROUND_LIGHT: Color = Color::rgb8(0x89, 0x89, 0x89);
pub const DISABLED_FOREGROUND_DARK: Color = Color::rgb8(0x6f, 0x6f, 0x6f);
pub const BUTTON_DARK: Color = Color::BLACK;
pub const BUTTON_LIGHT: Color = Color::rgb8(0x21, 0x21, 0x21);
pub const DISABLED_BUTTON_DARK: Color = Color::rgb8(0x28, 0x28, 0x28);
pub const DISABLED_BUTTON_LIGHT: Color = Color::rgb8(0x38, 0x38, 0x38);
pub const BUTTON_BORDER_RADIUS: f64 = 4.;
pub const BUTTON_BORDER_WIDTH: f64 = 2.;
pub const BORDER_DARK: Color = Color::rgb8(0x3a, 0x3a, 0x3a);
pub const BORDER_LIGHT: Color = Color::rgb8(0xa1, 0xa1, 0xa1);
pub const SELECTED_TEXT_BACKGROUND_COLOR: Color = Color::rgb8(0x43, 0x70, 0xA8);
pub const SELECTED_TEXT_INACTIVE_BACKGROUND_COLOR: Color = Color::rgb8(0x74, 0x74, 0x74);
pub const SELECTION_TEXT_COLOR: Color = Color::rgb8(0x00, 0x00, 0x00);
pub const CURSOR_COLOR: Color = Color::WHITE;
pub const TEXT_SIZE_NORMAL: f64 = 15.0;
pub const TEXT_SIZE_LARGE: f64 = 24.0;
pub const BASIC_WIDGET_HEIGHT: f64 = 18.0;
pub const WIDE_WIDGET_WIDTH: f64 = 100.;
pub const BORDERED_WIDGET_HEIGHT: f64 = 24.0;
pub const TEXTBOX_BORDER_RADIUS: f64 = 2.;
pub const TEXTBOX_BORDER_WIDTH: f64 = 1.;
pub const TEXTBOX_INSETS: Insets = Insets::new(4.0, 4.0, 4.0, 4.0);
pub const SCROLLBAR_COLOR: Color = Color::rgb8(0xff, 0xff, 0xff);
pub const SCROLLBAR_BORDER_COLOR: Color = Color::rgb8(0x77, 0x77, 0x77);
pub const SCROLLBAR_MAX_OPACITY: f64 = 0.7;
pub const SCROLLBAR_FADE_DELAY: u64 = 1500u64;
pub const SCROLLBAR_WIDTH: f64 = 8.;
pub const SCROLLBAR_PAD: f64 = 2.;
pub const SCROLLBAR_MIN_SIZE: f64 = 45.;
pub const SCROLLBAR_RADIUS: f64 = 5.;
pub const SCROLLBAR_EDGE_WIDTH: f64 = 1.;
pub const WIDGET_PADDING_VERTICAL: f64 = 10.0;
pub const WIDGET_PADDING_HORIZONTAL: f64 = 8.0;
pub const WIDGET_CONTROL_COMPONENT_PADDING: f64 = 4.0;

static DEBUG_COLOR: &[Color] = &[
    Color::rgb8(230, 25, 75),
    Color::rgb8(60, 180, 75),
    Color::rgb8(255, 225, 25),
    Color::rgb8(0, 130, 200),
    Color::rgb8(245, 130, 48),
    Color::rgb8(70, 240, 240),
    Color::rgb8(240, 50, 230),
    Color::rgb8(250, 190, 190),
    Color::rgb8(0, 128, 128),
    Color::rgb8(230, 190, 255),
    Color::rgb8(170, 110, 40),
    Color::rgb8(255, 250, 200),
    Color::rgb8(128, 0, 0),
    Color::rgb8(170, 255, 195),
    Color::rgb8(0, 0, 128),
    Color::rgb8(128, 128, 128),
    Color::rgb8(255, 255, 255),
    Color::rgb8(0, 0, 0),
];

pub fn get_debug_color(id: u64) -> Color {
    let color_num = id as usize % DEBUG_COLOR.len();
    DEBUG_COLOR[color_num]
}
