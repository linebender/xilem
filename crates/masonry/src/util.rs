// This software is licensed under Apache License 2.0 and distributed on an
// "as-is" basis without warranties of any kind. See the LICENSE file for
// details.

//! Miscellaneous utility functions.

#![cfg(not(tarpaulin_include))]

use std::any::Any;
use std::hash::Hash;

/// Panic in debug and tracing::error in release mode.
///
/// This macro is in some way a combination of `panic` and `debug_assert`,
/// but it will log the provided message instead of ignoring it in release builds.
///
/// It's useful when a backtrace would aid debugging but a crash can be avoided in release.
macro_rules! debug_panic {
    () => { ... };
    ($msg:expr) => {
        if cfg!(debug_assertions) {
            panic!($msg);
        } else {
            tracing::error!($msg);
        }
    };
    ($msg:expr,) => { debug_panic!($msg) };
    ($fmt:expr, $($arg:tt)+) => {
        if cfg!(debug_assertions) {
            panic!($fmt, $($arg)*);
        } else {
            tracing::error!($fmt, $($arg)*);
        }
    };
}

// ---

/// An enum for specifying whether an event was handled.
#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
pub enum Handled {
    /// An event was already handled, and shouldn't be propagated to other event handlers.
    Yes,
    /// An event has not yet been handled.
    No,
}

impl Handled {
    /// Has the event been handled yet?
    pub fn is_handled(self) -> bool {
        self == Handled::Yes
    }
}

impl From<bool> for Handled {
    /// Returns `Handled::Yes` if `handled` is true, and `Handled::No` otherwise.
    fn from(handled: bool) -> Handled {
        if handled {
            Handled::Yes
        } else {
            Handled::No
        }
    }
}

// ---

/// Trait extending Any, implemented for all types that implement Any.
///
/// This is a band-aid to substitute for a lack of dyn trait upcasting.
pub trait AsAny: Any {
    /// Return self.
    fn as_dyn_any(&self) -> &dyn Any;
    /// Return self.
    fn as_mut_dyn_any(&mut self) -> &mut dyn Any;
}

impl<T: Any> AsAny for T {
    fn as_dyn_any(&self) -> &dyn Any {
        self
    }

    fn as_mut_dyn_any(&mut self) -> &mut dyn Any {
        self
    }
}
