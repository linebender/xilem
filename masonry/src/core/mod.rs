// Copyright 2025 the Xilem Authors
// SPDX-License-Identifier: Apache-2.0

//! Basic types and traits Masonry is built on.

mod action;
mod box_constraints;
mod contexts;
mod event;
mod object_fit;
mod text;
#[allow(missing_docs, reason = "TODO")]
mod widget;
mod widget_arena;
#[allow(missing_docs, reason = "TODO")]
mod widget_mut;
#[allow(missing_docs, reason = "TODO")]
mod widget_pod;
mod widget_ref;
mod widget_state;

pub use action::Action;
pub use box_constraints::BoxConstraints;
pub use contexts::{
    AccessCtx, ComposeCtx, EventCtx, IsContext, LayoutCtx, MutateCtx, PaintCtx, QueryCtx,
    RawWrapper, RawWrapperMut, RegisterCtx, UpdateCtx,
};
pub use event::{
    AccessEvent, PointerButton, PointerEvent, PointerState, TextEvent, Update, WindowEvent,
    WindowTheme,
};
pub use object_fit::ObjectFit;
pub use text::{render_text, ArcStr, BrushIndex, StyleProperty, StyleSet};
pub use widget::find_widget_at_pos;
pub use widget::{AllowRawMut, FromDynWidget, Widget, WidgetId};
pub use widget_mut::WidgetMut;
pub use widget_pod::WidgetPod;
pub use widget_ref::WidgetRef;

pub(crate) use text::default_styles;
pub(crate) use widget_arena::WidgetArena;
pub(crate) use widget_pod::CreateWidget;
pub(crate) use widget_state::WidgetState;
