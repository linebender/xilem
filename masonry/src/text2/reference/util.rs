// Copyright 2018 the Xilem Authors and the Druid Authors
// SPDX-License-Identifier: Apache-2.0

use std::ops::{Bound, Range, RangeBounds};

/// Resolves a `RangeBounds` into a range in the range 0..len.
pub fn resolve_range(range: impl RangeBounds<usize>, len: usize) -> Range<usize> {
    let start = match range.start_bound() {
        Bound::Unbounded => 0,
        Bound::Included(n) => *n,
        Bound::Excluded(n) => *n + 1,
    };

    let end = match range.end_bound() {
        Bound::Unbounded => len,
        Bound::Included(n) => *n + 1,
        Bound::Excluded(n) => *n,
    };

    start.min(len)..end.min(len)
}
