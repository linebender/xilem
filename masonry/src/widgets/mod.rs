// Copyright 2018 the Xilem Authors and the Druid Authors
// SPDX-License-Identifier: Apache-2.0

//! Common widgets.

// We use allow because expect(missing_docs) is noisy with rust-analyzer.
#![allow(missing_docs, reason = "We have many as-yet undocumented items")]

#[cfg(test)]
mod tests;

mod align;
mod button;
mod checkbox;
mod flex;
mod grid;
mod image;
mod label;
mod portal;
mod progress_bar;
mod prose;
mod root_widget;
mod scroll_bar;
mod sized_box;
mod spinner;
mod split;
mod text_area;
mod textbox;
mod variable_label;
mod zstack;

pub use self::align::Align;
pub use self::button::Button;
pub use self::checkbox::Checkbox;
pub use self::flex::{Axis, CrossAxisAlignment, Flex, FlexParams, MainAxisAlignment};
pub use self::grid::{Grid, GridParams};
pub use self::image::Image;
pub use self::label::{Label, LineBreaking};
pub use self::portal::Portal;
pub use self::progress_bar::ProgressBar;
pub use self::prose::Prose;
pub use self::root_widget::RootWidget;
pub use self::scroll_bar::ScrollBar;
pub use self::sized_box::{Padding, SizedBox};
pub use self::spinner::Spinner;
pub use self::split::Split;
pub use self::text_area::TextArea;
pub use self::textbox::Textbox;
pub use self::variable_label::VariableLabel;
pub use self::zstack::{Alignment, ChildAlignment, HorizontalAlignment, VerticalAlignment, ZStack};
