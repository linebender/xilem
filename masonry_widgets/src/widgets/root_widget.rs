// Copyright 2024 the Xilem Authors
// SPDX-License-Identifier: Apache-2.0

use std::any::TypeId;

use accesskit::{Node, Role};
use smallvec::{SmallVec, smallvec};
use tracing::{Span, trace_span};
use vello::Scene;
use vello::kurbo::Point;

use crate::core::{
    AccessCtx, AccessEvent, BoxConstraints, EventCtx, LayoutCtx, PaintCtx, PointerEvent,
    PropertiesMut, PropertiesRef, QueryCtx, RegisterCtx, TextEvent, UpdateCtx, Widget, WidgetId,
    WidgetMut, WidgetPod,
};
use crate::kurbo::Size;
use crate::properties::{Background, Padding};
use crate::util::fill;

/// A wrapper Widget which app drivers can wrap around the rest of the widget tree.
///
/// This is useful for a few things:
/// - Reporting a [`Role::Window`] to the accessibility API.
/// - Setting a default [`Background`] and [`Padding`] for the entire app using [`DefaultProperties`].
///
/// [`DefaultProperties`]: crate::core::DefaultProperties
pub struct RootWidget {
    pub(crate) pod: WidgetPod<dyn Widget>,
}

impl RootWidget {
    /// Create a new root widget.
    pub fn new(widget: impl Widget) -> Self {
        Self {
            pod: WidgetPod::new(widget).erased(),
        }
    }

    /// Create a new root widget from a [`WidgetPod`].
    pub fn from_pod(pod: WidgetPod<dyn Widget>) -> Self {
        Self { pod }
    }
}

impl RootWidget {
    /// Get a mutable reference to the child widget.
    pub fn child_mut<'t>(this: &'t mut WidgetMut<'_, Self>) -> WidgetMut<'t, dyn Widget> {
        this.ctx.get_mut(&mut this.widget.pod)
    }
}

impl Widget for RootWidget {
    fn on_pointer_event(
        &mut self,
        _ctx: &mut EventCtx,
        _props: &mut PropertiesMut<'_>,
        _event: &PointerEvent,
    ) {
    }
    fn on_text_event(
        &mut self,
        _ctx: &mut EventCtx,
        _props: &mut PropertiesMut<'_>,
        _event: &TextEvent,
    ) {
    }
    fn on_access_event(
        &mut self,
        _ctx: &mut EventCtx,
        _props: &mut PropertiesMut<'_>,
        _event: &AccessEvent,
    ) {
    }

    fn register_children(&mut self, ctx: &mut RegisterCtx) {
        ctx.register_child(&mut self.pod);
    }

    fn property_changed(&mut self, ctx: &mut UpdateCtx, property_type: TypeId) {
        Background::prop_changed(ctx, property_type);
        Padding::prop_changed(ctx, property_type);
    }

    fn layout(
        &mut self,
        ctx: &mut LayoutCtx,
        props: &mut PropertiesMut<'_>,
        bc: &BoxConstraints,
    ) -> Size {
        let padding = props.get::<Padding>();

        let bc = padding.layout_down(*bc);
        let size = ctx.run_layout(&mut self.pod, &bc);
        let (size, _) = padding.layout_up(size, 0.);

        let pos = padding.place_down(Point::ORIGIN);
        ctx.place_child(&mut self.pod, pos);
        size
    }

    fn paint(&mut self, ctx: &mut PaintCtx, props: &PropertiesRef<'_>, scene: &mut Scene) {
        let bg = props.get::<Background>();
        let bg_rect = ctx.size().to_rect();
        let bg_brush = bg.get_peniko_brush_for_rect(bg_rect);

        fill(scene, &bg_rect, &bg_brush);
    }

    fn accessibility_role(&self) -> Role {
        Role::Window
    }

    fn accessibility(
        &mut self,
        _ctx: &mut AccessCtx,
        _props: &PropertiesRef<'_>,
        _node: &mut Node,
    ) {
    }

    fn children_ids(&self) -> SmallVec<[WidgetId; 16]> {
        smallvec![self.pod.id()]
    }

    fn make_trace_span(&self, ctx: &QueryCtx<'_>) -> Span {
        trace_span!("RootWidget", id = ctx.widget_id().trace())
    }
}
