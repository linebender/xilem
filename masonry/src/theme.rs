// Copyright 2019 the Xilem Authors and the Druid Authors
// SPDX-License-Identifier: Apache-2.0

//! Default values used by various widgets in their paint methods.

#![allow(missing_docs, reason = "Names are self-explanatory.")]

use parley::{GenericFamily, LineHeight};

use crate::core::{DefaultProperties, StyleProperty, StyleSet};
use crate::peniko::Color;
use crate::properties::{
    ActiveBackground, Background, BarColor, BorderColor, BorderWidth, CheckmarkColor,
    CheckmarkStrokeWidth, ContentColor, CornerRadius, DisabledBackground, DisabledCheckmarkColor,
    DisabledContentColor, HoveredBorderColor, Padding,
};
use crate::widgets::{Button, Checkbox, Label, ProgressBar, Spinner, TextArea, TextInput};

pub const BORDER_WIDTH: f64 = 1.;

// Zync color variations from https://tailwindcss.com/docs/colors
pub const ZYNC_900: Color = Color::from_rgb8(0x18, 0x18, 0x1b);
pub const ZYNC_800: Color = Color::from_rgb8(0x27, 0x27, 0x2a);
pub const ZYNC_700: Color = Color::from_rgb8(0x3f, 0x3f, 0x46);
pub const ZYNC_600: Color = Color::from_rgb8(0x52, 0x52, 0x5b);
pub const ZYNC_500: Color = Color::from_rgb8(0x71, 0x71, 0x7a);

pub const ACCENT_COLOR: Color = Color::from_rgb8(0x3b, 0x7e, 0xe4);
pub const TEXT_COLOR: Color = Color::from_rgb8(0xf0, 0xf0, 0xea);
pub const DISABLED_TEXT_COLOR: Color = Color::from_rgb8(0xa0, 0xa0, 0x9a);

/// Default horizontal padding for [`Label`], in logical pixels.
pub const LABEL_X_PADDING: f64 = 2.0;

// TODO: The following constants are not being used in properties
pub const TEXT_SIZE_NORMAL: f32 = 15.0;
pub const BASIC_WIDGET_HEIGHT: f64 = 18.0;
pub const BORDERED_WIDGET_HEIGHT: f64 = 24.0;
pub const SCROLLBAR_COLOR: Color = Color::from_rgb8(0xff, 0xff, 0xff);
pub const SCROLLBAR_BORDER_COLOR: Color = Color::from_rgb8(0x77, 0x77, 0x77);
pub const SCROLLBAR_WIDTH: f64 = 8.;
pub const SCROLLBAR_PAD: f64 = 2.;
pub const SCROLLBAR_MIN_SIZE: f64 = 45.;
pub const SCROLLBAR_RADIUS: f64 = 5.;
pub const SCROLLBAR_EDGE_WIDTH: f64 = 1.;
pub const DEFAULT_GAP: f64 = 10.0;
pub const DEFAULT_SPACER_LEN: f64 = 10.0;
pub const WIDGET_CONTROL_COMPONENT_PADDING: f64 = 4.0;

pub fn default_property_set() -> DefaultProperties {
    let mut properties = DefaultProperties::new();

    // Button
    properties.insert::<Button, _>(Padding::from_vh(6., 16.));
    properties.insert::<Button, _>(CornerRadius { radius: 6. });
    properties.insert::<Button, _>(BorderWidth {
        width: BORDER_WIDTH,
    });

    properties.insert::<Button, _>(Background::Color(ZYNC_800));
    properties.insert::<Button, _>(ActiveBackground(Background::Color(ZYNC_700)));
    properties.insert::<Button, _>(DisabledBackground(Background::Color(Color::BLACK)));
    properties.insert::<Button, _>(BorderColor { color: ZYNC_700 });
    properties.insert::<Button, _>(HoveredBorderColor(BorderColor { color: ZYNC_500 }));

    // Checkbox
    properties.insert::<Checkbox, _>(CornerRadius { radius: 4. });
    properties.insert::<Checkbox, _>(BorderWidth {
        width: BORDER_WIDTH,
    });

    properties.insert::<Checkbox, _>(Background::Color(ZYNC_800));
    properties.insert::<Checkbox, _>(ActiveBackground(Background::Color(ZYNC_700)));
    properties.insert::<Checkbox, _>(DisabledBackground(Background::Color(Color::BLACK)));
    properties.insert::<Checkbox, _>(BorderColor { color: ZYNC_700 });
    properties.insert::<Checkbox, _>(HoveredBorderColor(BorderColor { color: ZYNC_500 }));

    properties.insert::<Checkbox, _>(CheckmarkStrokeWidth { width: 2.0 });
    properties.insert::<Checkbox, _>(CheckmarkColor { color: TEXT_COLOR });
    properties.insert::<Checkbox, _>(DisabledCheckmarkColor(CheckmarkColor {
        color: DISABLED_TEXT_COLOR,
    }));

    // TextInput
    properties.insert::<TextInput, _>(Padding::from_vh(6., 12.));
    properties.insert::<TextInput, _>(CornerRadius { radius: 4. });
    properties.insert::<TextInput, _>(BorderWidth {
        width: BORDER_WIDTH,
    });

    properties.insert::<TextInput, _>(BorderColor { color: ZYNC_600 });

    // TextArea
    properties.insert::<TextArea<false>, _>(ContentColor::new(TEXT_COLOR));
    properties
        .insert::<TextArea<false>, _>(DisabledContentColor(ContentColor::new(DISABLED_TEXT_COLOR)));
    properties.insert::<TextArea<true>, _>(ContentColor::new(TEXT_COLOR));
    properties
        .insert::<TextArea<true>, _>(DisabledContentColor(ContentColor::new(DISABLED_TEXT_COLOR)));

    // Label
    properties.insert::<Label, _>(Padding::from_vh(0., LABEL_X_PADDING));
    properties.insert::<Label, _>(ContentColor::new(TEXT_COLOR));
    properties.insert::<Label, _>(DisabledContentColor(ContentColor::new(DISABLED_TEXT_COLOR)));

    // ProgressBar
    properties.insert::<ProgressBar, _>(CornerRadius { radius: 2. });
    properties.insert::<ProgressBar, _>(BorderWidth {
        width: BORDER_WIDTH,
    });

    properties.insert::<ProgressBar, _>(Background::Color(ZYNC_900));
    properties.insert::<ProgressBar, _>(BorderColor { color: ZYNC_800 });
    properties.insert::<ProgressBar, _>(BarColor(ACCENT_COLOR));

    // Spinner
    properties.insert::<Spinner, _>(ContentColor::new(TEXT_COLOR));

    properties
}

/// Applies the default text styles for Masonry into `styles`.
pub fn default_text_styles(styles: &mut StyleSet) {
    styles.insert(StyleProperty::LineHeight(LineHeight::FontSizeRelative(1.2)));
    styles.insert(GenericFamily::SystemUi.into());
}
