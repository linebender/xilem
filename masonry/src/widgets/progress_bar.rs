// Copyright 2019 the Xilem Authors and the Druid Authors
// SPDX-License-Identifier: Apache-2.0

//! A progress bar widget.

use accesskit::{Node, Role};
use smallvec::{SmallVec, smallvec};
use tracing::{Span, trace_span};
use vello::Scene;

use crate::core::{
    AccessCtx, AccessEvent, ArcStr, BoxConstraints, EventCtx, LayoutCtx, PaintCtx, PointerEvent,
    PropertiesMut, PropertiesRef, QueryCtx, RegisterCtx, TextEvent, Update, UpdateCtx, Widget,
    WidgetId, WidgetMut, WidgetPod,
};
use crate::kurbo::{Point, Size};
use crate::theme;
use crate::util::{UnitPoint, fill_lin_gradient, stroke};
use crate::widgets::{Label, LineBreaking};

/// A progress bar.
///
#[doc = crate::include_screenshot!("widget/screenshots/masonry__widget__progress_bar__tests__25_percent_progressbar.png", "25% progress bar.")]
pub struct ProgressBar {
    /// A value in the range `[0, 1]` inclusive, where 0 is 0% and 1 is 100% complete.
    ///
    /// `None` variant can be used to show a progress bar without a percentage.
    /// It is also used if an invalid float (outside of [0, 1]) is passed.
    progress: Option<f64>,
    label: WidgetPod<Label>,
}

impl ProgressBar {
    /// Create a new `ProgressBar`.
    ///
    /// `progress` is a number between 0 and 1 inclusive. If it is `NaN`, then an
    /// indefinite progress bar will be shown.
    /// Otherwise, the input will be clamped to [0, 1].
    pub fn new(mut progress: Option<f64>) -> Self {
        clamp_progress(&mut progress);
        let label = WidgetPod::new(
            Label::new(Self::value(progress)).with_line_break_mode(LineBreaking::Overflow),
        );
        Self { progress, label }
    }

    fn value_accessibility(&self) -> Box<str> {
        if let Some(value) = self.progress {
            format!("{:.0}%", value * 100.).into()
        } else {
            "progress unspecified".into()
        }
    }

    fn value(progress: Option<f64>) -> ArcStr {
        if let Some(value) = progress {
            format!("{:.0}%", value * 100.).into()
        } else {
            "".into()
        }
    }
}

// --- MARK: WIDGETMUT ---
impl ProgressBar {
    pub fn set_progress(this: &mut WidgetMut<'_, Self>, mut progress: Option<f64>) {
        clamp_progress(&mut progress);
        let progress_changed = this.widget.progress != progress;
        if progress_changed {
            this.widget.progress = progress;
            let mut label = this.ctx.get_mut(&mut this.widget.label);
            Label::set_text(&mut label, Self::value(progress));
        }
        this.ctx.request_layout();
        this.ctx.request_render();
    }
}

/// Helper to ensure progress is either a number between [0, 1] inclusive, or `None`.
///
/// NaNs are converted to `None`.
fn clamp_progress(progress: &mut Option<f64>) {
    if let Some(value) = progress {
        if value.is_nan() {
            *progress = None;
        } else {
            *progress = Some(value.clamp(0., 1.));
        }
    }
}

// --- MARK: IMPL WIDGET ---
impl Widget for ProgressBar {
    fn on_pointer_event(
        &mut self,
        _ctx: &mut EventCtx,
        _props: &mut PropertiesMut<'_>,
        _event: &PointerEvent,
    ) {
    }

    fn on_text_event(
        &mut self,
        _ctx: &mut EventCtx,
        _props: &mut PropertiesMut<'_>,
        _event: &TextEvent,
    ) {
    }

    fn on_access_event(
        &mut self,
        _ctx: &mut EventCtx,
        _props: &mut PropertiesMut<'_>,
        _event: &AccessEvent,
    ) {
    }

    fn register_children(&mut self, ctx: &mut RegisterCtx) {
        ctx.register_child(&mut self.label);
    }

    fn update(&mut self, _ctx: &mut UpdateCtx, _props: &mut PropertiesMut<'_>, _event: &Update) {}

    fn layout(
        &mut self,
        ctx: &mut LayoutCtx,
        _props: &mut PropertiesMut<'_>,
        bc: &BoxConstraints,
    ) -> Size {
        const DEFAULT_WIDTH: f64 = 400.;
        // TODO: Clearer constraints here
        let label_size = ctx.run_layout(&mut self.label, &bc.loosen());
        let desired_size = Size::new(
            DEFAULT_WIDTH.max(label_size.width),
            crate::theme::BASIC_WIDGET_HEIGHT.max(label_size.height),
        );
        let final_size = bc.constrain(desired_size);

        // center text
        let text_pos = Point::new(
            ((final_size.width - label_size.width) * 0.5).max(0.),
            ((final_size.height - label_size.height) * 0.5).max(0.),
        );
        ctx.place_child(&mut self.label, text_pos);
        final_size
    }

    fn paint(&mut self, ctx: &mut PaintCtx, _props: &PropertiesRef<'_>, scene: &mut Scene) {
        let border_width = 1.;
        let size = ctx.size();
        let border_radius = 2.;

        let bg_rect = size
            .to_rect()
            .inset(-border_width)
            .to_rounded_rect(border_radius - border_width);
        let border_rect = size
            .to_rect()
            .inset(-border_width / 2.0)
            .to_rounded_rect(border_radius);

        let progress_rect_size = Size::new(
            ctx.size().width * self.progress.unwrap_or(1.),
            ctx.size().height,
        );
        let progress_rect = progress_rect_size
            .to_rect()
            .inset(-border_width)
            .to_rounded_rect(border_radius - border_width);

        fill_lin_gradient(
            scene,
            &bg_rect,
            [theme::BACKGROUND_LIGHT, theme::BACKGROUND_DARK],
            UnitPoint::TOP,
            UnitPoint::BOTTOM,
        );
        fill_lin_gradient(
            scene,
            &progress_rect,
            [theme::PRIMARY_LIGHT, theme::PRIMARY_DARK],
            UnitPoint::TOP,
            UnitPoint::BOTTOM,
        );
        stroke(scene, &border_rect, theme::BORDER_DARK, border_width);
    }

    fn accessibility_role(&self) -> Role {
        Role::ProgressIndicator
    }

    fn accessibility(&mut self, _ctx: &mut AccessCtx, _props: &PropertiesRef<'_>, node: &mut Node) {
        node.set_value(self.value_accessibility());
        if let Some(value) = self.progress {
            node.set_numeric_value(value * 100.0);
        }
    }

    fn children_ids(&self) -> SmallVec<[WidgetId; 16]> {
        smallvec![self.label.id()]
    }

    fn make_trace_span(&self, ctx: &QueryCtx<'_>) -> Span {
        trace_span!("ProgressBar", id = ctx.widget_id().trace())
    }

    fn get_debug_text(&self) -> Option<String> {
        Some(self.value_accessibility().into())
    }
}

// --- MARK: TESTS ---
#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;

    use super::*;
    use crate::assert_render_snapshot;
    use crate::testing::{TestHarness, TestWidgetExt, widget_ids};

    #[test]
    fn indeterminate_progressbar() {
        let [progressbar_id] = widget_ids();
        let widget = ProgressBar::new(None).with_id(progressbar_id);

        let mut harness = TestHarness::create(widget);

        assert_debug_snapshot!(harness.root_widget());
        assert_render_snapshot!(harness, "indeterminate_progressbar");
    }

    #[test]
    fn _0_percent_progressbar() {
        let [_0percent] = widget_ids();

        let widget = ProgressBar::new(Some(0.)).with_id(_0percent);
        let mut harness = TestHarness::create(widget);
        assert_debug_snapshot!(harness.root_widget());
        assert_render_snapshot!(harness, "0_percent_progressbar");
    }

    #[test]
    fn _25_percent_progressbar() {
        let [_25percent] = widget_ids();

        let widget = ProgressBar::new(Some(0.25)).with_id(_25percent);
        let mut harness = TestHarness::create(widget);
        assert_debug_snapshot!(harness.root_widget());
        assert_render_snapshot!(harness, "25_percent_progressbar");
    }

    #[test]
    fn _50_percent_progressbar() {
        let [_50percent] = widget_ids();

        let widget = ProgressBar::new(Some(0.5)).with_id(_50percent);
        let mut harness = TestHarness::create(widget);
        assert_debug_snapshot!(harness.root_widget());
        assert_render_snapshot!(harness, "50_percent_progressbar");
    }

    #[test]
    fn _75_percent_progressbar() {
        let [_75percent] = widget_ids();

        let widget = ProgressBar::new(Some(0.75)).with_id(_75percent);
        let mut harness = TestHarness::create(widget);
        assert_debug_snapshot!(harness.root_widget());
        assert_render_snapshot!(harness, "75_percent_progressbar");
    }

    #[test]
    fn _100_percent_progressbar() {
        let [_100percent] = widget_ids();

        let widget = ProgressBar::new(Some(1.)).with_id(_100percent);
        let mut harness = TestHarness::create(widget);
        assert_debug_snapshot!(harness.root_widget());
        assert_render_snapshot!(harness, "100_percent_progressbar");
    }

    #[test]
    fn edit_progressbar() {
        let image_1 = {
            let bar = ProgressBar::new(Some(0.5));

            let mut harness = TestHarness::create_with_size(bar, Size::new(60.0, 20.0));

            harness.render()
        };

        let image_2 = {
            let bar = ProgressBar::new(None);

            let mut harness = TestHarness::create_with_size(bar, Size::new(60.0, 20.0));

            harness.edit_root_widget(|mut label| {
                let mut bar = label.downcast::<ProgressBar>();
                ProgressBar::set_progress(&mut bar, Some(0.5));
            });

            harness.render()
        };

        // We don't use assert_eq because we don't want rich assert
        assert!(image_1 == image_2);
    }
}
