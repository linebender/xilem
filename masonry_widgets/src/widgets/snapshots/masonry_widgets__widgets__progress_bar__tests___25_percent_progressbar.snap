---
source: masonry/src/widgets/progress_bar.rs
assertion_line: 258
expression: harness.root_widget()
---
WrapperWidget(
    ProgressBar<25%>(
        Label<25%>,
    ),
)
